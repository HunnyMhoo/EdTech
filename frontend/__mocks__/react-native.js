const ReactNative = require('react-native');

ReactNative.StyleSheet = {
  create: styles => styles,
};

module.exports = ReactNative; 