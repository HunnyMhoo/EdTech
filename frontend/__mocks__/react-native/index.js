const RN = jest.requireActual('react-native');

RN.StyleSheet = {
  create: (styles) => styles,
};

module.exports = RN; 