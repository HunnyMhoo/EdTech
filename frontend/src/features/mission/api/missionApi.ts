import { API_BASE_URL } from '@/config';

// Enhanced answer interface with attempt tracking
export interface AnswerAttempt {
  answer: any;
  is_correct: boolean;
  timestamp: string;
}

export interface Answer {
  question_id: string;
  current_answer: any;
  is_correct: boolean;
  attempt_count: number;
  attempts_history: AnswerAttempt[];
  feedback_shown: boolean;
  is_complete: boolean;
  max_retries: number;
}

// Legacy answer interface for backward compatibility
export interface LegacyAnswer {
  question_id: string;
  answer: any;
  feedback_shown?: boolean;
}

// Define the Question type based on the backend model
export interface Question {
  question_id: string;
  question_text: string;
  skill_area: string;
  difficulty_level: number;
  choices?: Array<{ id: string; text: string }>;
  feedback_th: string;
  correct_answer_id: string;
}

// Updated Mission interface to match DailyMissionDocument from backend
export interface Mission {
  user_id: string;
  date: string;
  questions: Question[];
  status: 'not_started' | 'in_progress' | 'complete' | 'archived';
  current_question_index: number;
  answers: Answer[];
  created_at: string;
  updated_at: string;
}

// Feedback response interface
export interface FeedbackResponse {
  already_complete: boolean;
  is_correct: boolean;
  correct_answer: string;
  explanation: string;
  attempt_count: number;
  max_retries?: number;
  can_retry?: boolean;
  question_complete?: boolean;
}

// API response wrapper
export interface ApiResponse<T = any> {
  status: string;
  message: string;
  data?: T;
  feedback?: FeedbackResponse;
}

// Payload for updating mission progress
export interface MissionProgressUpdatePayload {
  current_question_index: number;
  answers: LegacyAnswer[]; // Keep legacy format for backward compatibility
  status?: 'not_started' | 'in_progress' | 'complete' | 'archived';
}

// Payload for submitting answers
export interface AnswerSubmissionPayload {
  question_id: string;
  answer: any;
}

export const fetchDailyMission = async (userId: string): Promise<Mission> => {
  const response = await fetch(`${API_BASE_URL}/missions/daily/${userId}`, {
    method: 'GET',
    headers: {
      'Content-Type': 'application/json',
    },
  });

  if (!response.ok) {
    const errorData = await response.json();
    throw new Error(errorData.detail || `HTTP error! status: ${response.status}`);
  }

  const responseData: ApiResponse<Mission> = await response.json();
  return responseData.data!;
};

export const updateMissionProgressApi = async (
  userId: string,
  payload: MissionProgressUpdatePayload
): Promise<Mission> => {
  const response = await fetch(`${API_BASE_URL}/missions/daily/${userId}/progress`, {
    method: 'PUT',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    const errorData = await response.json();
    throw new Error(errorData.detail || `HTTP error! status: ${response.status}`);
  }

  const responseData: ApiResponse<Mission> = await response.json();
  return responseData.data!;
};

export const submitAnswerWithFeedback = async (
  userId: string,
  questionId: string,
  answer: any
): Promise<FeedbackResponse> => {
  const payload: AnswerSubmissionPayload = {
    question_id: questionId,
    answer: answer,
  };

  const response = await fetch(`${API_BASE_URL}/missions/daily/${userId}/submit-answer`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    const errorData = await response.json();
    throw new Error(errorData.detail || `HTTP error! status: ${response.status}`);
  }

  const responseData: ApiResponse = await response.json();
  return responseData.feedback!;
};

export const markFeedbackShown = async (
  userId: string,
  questionId: string
): Promise<{ mission_status: string }> => {
  const response = await fetch(`${API_BASE_URL}/missions/daily/${userId}/mark-feedback-shown`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ question_id: questionId }),
  });

  if (!response.ok) {
    const errorData = await response.json();
    throw new Error(errorData.detail || `HTTP error! status: ${response.status}`);
  }

  const responseData: any = await response.json();
  return { mission_status: responseData.mission_status || 'in_progress' };
};

export const retryQuestion = async (
  userId: string,
  questionId: string
): Promise<{ remaining_attempts: number }> => {
  const response = await fetch(`${API_BASE_URL}/missions/daily/${userId}/retry-question`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ question_id: questionId }),
  });

  if (!response.ok) {
    const errorData = await response.json();
    throw new Error(errorData.detail || `HTTP error! status: ${response.status}`);
  }

  const responseData: any = await response.json();
  return { remaining_attempts: responseData.remaining_attempts || 0 };
};

// --- Mocked Data (can be removed or kept for testing if backend is down) ---
// Example: To use mock data if API_BASE_URL is not reachable or for specific tests.
const MOCK_MODE = false; // Set to true to use mock data unconditionally

const getMockMission = async (): Promise<Mission> => {
    console.log('Using mocked mission data (getMockMission)');
    await new Promise(resolve => setTimeout(resolve, 300));
    return {
      user_id: 'mock_user_123',
      date: new Date().toISOString().split('T')[0],
      questions: ['q1_mock', 'q2_mock', 'q3_mock', 'q4_mock', 'q5_mock'].map((id, index) => ({
        question_id: id,
        question_text: `This is mock question ${index + 1}?`,
        skill_area: 'Mock',
        difficulty_level: 1,
        choices: [
            { id: `${id}_choice1`, text: 'Choice A' },
            { id: `${id}_choice2`, text: 'Choice B' },
        ],
        feedback_th: `นี่คือคำอธิบายจำลองสำหรับ ${id}`,
        correct_answer_id: `${id}_choice1`,
      })),
      status: 'not_started',
      current_question_index: 0,
      answers: [],
      created_at: new Date().toISOString(),
      updated_at: new Date().toISOString(),
    };
};

// You could modify fetchDailyMission and updateMissionProgressApi to use MOCK_MODE
// For example:
// if (MOCK_MODE) return getMockMission();
// inside fetchDailyMission before the try block. 