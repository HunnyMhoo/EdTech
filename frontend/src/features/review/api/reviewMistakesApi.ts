import { API_BASE_URL } from '@/config';

// API response wrapper (matching existing pattern)
export interface ApiResponse<T = any> {
  status: string;
  message?: string;
  data?: T;
}

// Types for Review Mistakes API
export interface ReviewMistakeItem {
  question_id: string;
  question_text: string;
  skill_area: string;
  difficulty_level: number;
  choices: Array<{
    id: string;
    text: string;
  }>;
  user_answer_id: string;
  user_answer_text: string;
  correct_answer_id: string;
  correct_answer_text: string;
  explanation: string;
  mission_date: string;  // ISO date string
  mission_completion_date: string;  // ISO datetime string
  attempt_count: number;
}

export interface PaginationInfo {
  current_page: number;
  total_pages: number;
  total_items: number;
  items_per_page: number;
  has_next: boolean;
  has_previous: boolean;
}

export interface ReviewMistakesResponse {
  mistakes: ReviewMistakeItem[];
  pagination: PaginationInfo;
  total_mistakes: number;
}

export interface GroupedReviewMistakesResponse {
  grouped_mistakes: Record<string, ReviewMistakeItem[]>;
  pagination: PaginationInfo;
  total_mistakes: number;
  group_counts: Record<string, number>;
}

export interface ReviewStatsResponse {
  total_mistakes: number;
  skill_areas_count: number;
  skill_areas: string[];
  difficulty_breakdown: Record<number, number>;
}

// API parameters
export interface ReviewMistakesParams {
  page?: number;
  items_per_page?: number;
  skill_area?: string;
}

export interface GroupedReviewMistakesParams {
  group_by: 'date' | 'topic';
  page?: number;
  items_per_page?: number;
  skill_area?: string;
}

// API functions following the pattern of missionApi.ts
export const getReviewMistakes = async (
  userId: string,
  params: ReviewMistakesParams = {}
): Promise<ReviewMistakesResponse> => {
  const queryParams = new URLSearchParams();
  
  if (params.page) queryParams.append('page', params.page.toString());
  if (params.items_per_page) queryParams.append('items_per_page', params.items_per_page.toString());
  if (params.skill_area) queryParams.append('skill_area', params.skill_area);

  const url = queryParams.toString() 
    ? `${API_BASE_URL}/review-mistakes/${userId}?${queryParams}` 
    : `${API_BASE_URL}/review-mistakes/${userId}`;
  
  const response = await fetch(url, {
    method: 'GET',
    headers: {
      'Content-Type': 'application/json',
    },
  });

  if (!response.ok) {
    const errorData = await response.json();
    throw new Error(errorData.detail || `HTTP error! status: ${response.status}`);
  }

  const responseData: ApiResponse<ReviewMistakesResponse> = await response.json();
  return responseData.data!;
};

export const getGroupedReviewMistakes = async (
  userId: string,
  params: GroupedReviewMistakesParams
): Promise<GroupedReviewMistakesResponse> => {
  const queryParams = new URLSearchParams();
  
  queryParams.append('group_by', params.group_by);
  if (params.page) queryParams.append('page', params.page.toString());
  if (params.items_per_page) queryParams.append('items_per_page', params.items_per_page.toString());
  if (params.skill_area) queryParams.append('skill_area', params.skill_area);

  const url = `${API_BASE_URL}/review-mistakes/${userId}/grouped?${queryParams}`;
  
  const response = await fetch(url, {
    method: 'GET',
    headers: {
      'Content-Type': 'application/json',
    },
  });

  if (!response.ok) {
    const errorData = await response.json();
    throw new Error(errorData.detail || `HTTP error! status: ${response.status}`);
  }

  const responseData: ApiResponse<GroupedReviewMistakesResponse> = await response.json();
  return responseData.data!;
};

export const getAvailableSkillAreas = async (userId: string): Promise<string[]> => {
  const url = `${API_BASE_URL}/review-mistakes/${userId}/skill-areas`;
  
  const response = await fetch(url, {
    method: 'GET',
    headers: {
      'Content-Type': 'application/json',
    },
  });

  if (!response.ok) {
    const errorData = await response.json();
    throw new Error(errorData.detail || `HTTP error! status: ${response.status}`);
  }

  const responseData: ApiResponse<string[]> = await response.json();
  return responseData.data!;
};

export const getReviewStats = async (userId: string): Promise<ReviewStatsResponse> => {
  const url = `${API_BASE_URL}/review-mistakes/${userId}/stats`;
  
  const response = await fetch(url, {
    method: 'GET',
    headers: {
      'Content-Type': 'application/json',
    },
  });

  if (!response.ok) {
    const errorData = await response.json();
    throw new Error(errorData.detail || `HTTP error! status: ${response.status}`);
  }

  const responseData: ApiResponse<ReviewStatsResponse> = await response.json();
  return responseData.data!;
}; 