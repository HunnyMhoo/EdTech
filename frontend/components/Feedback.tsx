import React from 'react';
import { View, Text, StyleSheet } from 'react-native';
import Toast, { BaseToast, ErrorToast } from 'react-native-toast-message';

const styles = StyleSheet.create({
  // Add any custom styles here if needed
  successToast: {
    borderLeftColor: 'green',
  },
  errorToast: {
    borderLeftColor: 'red',
  },
  contentContainer: {
    paddingHorizontal: 15,
  },
  title: {
    fontWeight: 'bold',
  },
  message: {
    fontSize: 14,
    color: '#333',
  },
});


/*
  1. Create the config
*/
export const toastConfig = {
  /*
    Overwrite 'success' type,
    by modifying the existing `BaseToast` component
  */
  success: (props: any) => (
    <BaseToast
      {...props}
      style={styles.successToast}
      contentContainerStyle={styles.contentContainer}
      text1Style={styles.title}
      text2Style={styles.message}
    />
  ),
  /*
    Overwrite 'error' type,
    by modifying the existing `ErrorToast` component
  */
  error: (props: any) => (
    <ErrorToast
      {...props}
      style={styles.errorToast}
      contentContainerStyle={styles.contentContainer}
      text1Style={styles.title}
      text2Style={styles.message}
    />
  ),
  /*
    Or create a completely new type - `tomatoToast`,
    building the layout from scratch.

    I've commented this out as we don't need it for now.
  */
  // tomatoToast: ({ text1, props }) => (
  //   <View style={{ height: 60, width: '100%', backgroundColor: 'tomato' }}>
  //     <Text>{text1}</Text>
  //     <Text>{props.uuid}</Text>
  //   </View>
  // )
}; 