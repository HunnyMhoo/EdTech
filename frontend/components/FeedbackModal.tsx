import React from 'react';
import { Modal, View, Text, StyleSheet, TouchableOpacity } from 'react-native';

interface FeedbackModalProps {
  isVisible: boolean;
  isCorrect: boolean;
  correctAnswer: string;
  explanation: string;
  onDismiss: () => void;
}

const FeedbackModal: React.FC<FeedbackModalProps> = ({
  isVisible,
  isCorrect,
  correctAnswer,
  explanation,
  onDismiss,
}) => {
  return (
    <Modal
      animationType="slide"
      transparent={true}
      visible={isVisible}
      onRequestClose={onDismiss}
    >
      <View style={styles.centeredView}>
        <View style={styles.modalView}>
          <Text style={[styles.title, { color: isCorrect ? 'green' : 'red' }]}>
            {isCorrect ? 'Correct!' : 'Incorrect'}
          </Text>
          <Text style={styles.text}>The correct answer is: {correctAnswer}</Text>
          <Text style={styles.text}>{explanation}</Text>
          <TouchableOpacity
            style={[styles.button, styles.buttonClose]}
            onPress={onDismiss}
          >
            <Text style={styles.textStyle}>Continue</Text>
          </TouchableOpacity>
        </View>
      </View>
    </Modal>
  );
};

const styles = StyleSheet.create({
  centeredView: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'rgba(0, 0, 0, 0.5)',
  },
  modalView: {
    margin: 20,
    backgroundColor: 'white',
    borderRadius: 20,
    padding: 35,
    alignItems: 'center',
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 2,
    },
    shadowOpacity: 0.25,
    shadowRadius: 4,
    elevation: 5,
  },
  title: {
    marginBottom: 15,
    textAlign: 'center',
    fontSize: 24,
    fontWeight: 'bold',
  },
  text: {
    marginBottom: 15,
    textAlign: 'center',
  },
  button: {
    borderRadius: 20,
    padding: 10,
    elevation: 2,
    minWidth: 100,
  },
  buttonClose: {
    backgroundColor: '#2196F3',
  },
  textStyle: {
    color: 'white',
    fontWeight: 'bold',
    textAlign: 'center',
  },
});

export default FeedbackModal; 