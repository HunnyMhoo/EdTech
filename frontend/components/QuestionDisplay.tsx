import React from 'react';
import { Text, StyleSheet } from 'react-native';

type QuestionDisplayProps = {
  questionIndex: number;
  questionText: string;
};

const QuestionDisplay: React.FC<QuestionDisplayProps> = ({ questionIndex, questionText }) => {
  return (
    <Text style={styles.questionText}>
      {`Question ${questionIndex + 1}: ${questionText}`}
    </Text>
  );
};

const styles = StyleSheet.create({
  questionText: {
    fontSize: 18,
    fontWeight: '500',
    marginBottom: 15,
  },
});

export default QuestionDisplay; 