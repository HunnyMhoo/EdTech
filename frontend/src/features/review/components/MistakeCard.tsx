import React from 'react';
import { View, Text, StyleSheet, TouchableOpacity } from 'react-native';
import { ReviewMistakeItem } from '../api/reviewMistakesApi';

interface MistakeCardProps {
  mistake: ReviewMistakeItem;
  onPress: () => void;
  showDate?: boolean;
  compact?: boolean;
}

export const MistakeCard: React.FC<MistakeCardProps> = ({ 
  mistake, 
  onPress, 
  showDate = true,
  compact = false 
}) => {
  const formatDate = (dateString: string) => {
    try {
      const date = new Date(dateString);
      return date.toLocaleDateString('th-TH', {
        day: '2-digit',
        month: 'short',
        year: 'numeric'
      });
    } catch {
      return dateString;
    }
  };

  const getDifficultyColor = (level: number) => {
    switch (level) {
      case 1: return '#22C55E'; // Green for easy
      case 2: return '#F59E0B'; // Orange for medium
      case 3: return '#EF4444'; // Red for hard
      default: return '#6B7280'; // Gray for unknown
    }
  };

  const getDifficultyText = (level: number) => {
    switch (level) {
      case 1: return 'Easy';
      case 2: return 'Medium';
      case 3: return 'Hard';
      default: return `Level ${level}`;
    }
  };

  return (
    <TouchableOpacity 
      style={[styles.card, compact && styles.compactCard]} 
      onPress={onPress}
      activeOpacity={0.7}
    >
      {/* Header with skill area and difficulty */}
      <View style={styles.header}>
        <View style={styles.skillAreaContainer}>
          <Text style={styles.skillArea}>{mistake.skill_area}</Text>
          <View style={[
            styles.difficultyBadge, 
            { backgroundColor: getDifficultyColor(mistake.difficulty_level) }
          ]}>
            <Text style={styles.difficultyText}>
              {getDifficultyText(mistake.difficulty_level)}
            </Text>
          </View>
        </View>
        
        {showDate && (
          <Text style={styles.date}>
            {formatDate(mistake.mission_date)}
          </Text>
        )}
      </View>

      {/* Question text */}
      <Text 
        style={[styles.questionText, compact && styles.compactQuestionText]}
        numberOfLines={compact ? 2 : 3}
      >
        {mistake.question_text}
      </Text>

      {/* Answer comparison */}
      <View style={styles.answerSection}>
        <View style={styles.answerRow}>
          <View style={styles.answerLabel}>
            <View style={styles.incorrectIndicator} />
            <Text style={styles.answerLabelText}>Your answer:</Text>
          </View>
          <Text style={[styles.answerText, styles.incorrectAnswer]} numberOfLines={1}>
            {mistake.user_answer_text}
          </Text>
        </View>
        
        <View style={styles.answerRow}>
          <View style={styles.answerLabel}>
            <View style={styles.correctIndicator} />
            <Text style={styles.answerLabelText}>Correct answer:</Text>
          </View>
          <Text style={[styles.answerText, styles.correctAnswer]} numberOfLines={1}>
            {mistake.correct_answer_text}
          </Text>
        </View>
      </View>

      {/* Footer with attempt count */}
      <View style={styles.footer}>
        <Text style={styles.attemptCount}>
          {mistake.attempt_count} attempt{mistake.attempt_count !== 1 ? 's' : ''}
        </Text>
        <View style={styles.viewDetailsContainer}>
          <Text style={styles.viewDetailsText}>Tap to view details</Text>
          <Text style={styles.chevron}>â€º</Text>
        </View>
      </View>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  card: {
    backgroundColor: '#FFFFFF',
    borderRadius: 12,
    padding: 16,
    marginBottom: 12,
    shadowColor: '#000000',
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.1,
    shadowRadius: 8,
    elevation: 3,
    borderWidth: 1,
    borderColor: '#F3F4F6',
  },
  compactCard: {
    padding: 12,
    marginBottom: 8,
  },
  header: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'flex-start',
    marginBottom: 12,
  },
  skillAreaContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    flex: 1,
  },
  skillArea: {
    fontSize: 14,
    fontWeight: '600',
    color: '#374151',
    marginRight: 8,
  },
  difficultyBadge: {
    paddingHorizontal: 8,
    paddingVertical: 4,
    borderRadius: 12,
  },
  difficultyText: {
    fontSize: 12,
    fontWeight: '500',
    color: '#FFFFFF',
  },
  date: {
    fontSize: 12,
    color: '#6B7280',
    fontWeight: '500',
  },
  questionText: {
    fontSize: 16,
    fontWeight: '500',
    color: '#111827',
    lineHeight: 22,
    marginBottom: 16,
  },
  compactQuestionText: {
    fontSize: 14,
    marginBottom: 12,
  },
  answerSection: {
    marginBottom: 16,
  },
  answerRow: {
    flexDirection: 'row',
    alignItems: 'center',
    marginBottom: 8,
  },
  answerLabel: {
    flexDirection: 'row',
    alignItems: 'center',
    width: 120,
  },
  incorrectIndicator: {
    width: 8,
    height: 8,
    borderRadius: 4,
    backgroundColor: '#EF4444',
    marginRight: 8,
  },
  correctIndicator: {
    width: 8,
    height: 8,
    borderRadius: 4,
    backgroundColor: '#22C55E',
    marginRight: 8,
  },
  answerLabelText: {
    fontSize: 12,
    color: '#6B7280',
    fontWeight: '500',
  },
  answerText: {
    fontSize: 14,
    fontWeight: '500',
    flex: 1,
  },
  incorrectAnswer: {
    color: '#EF4444',
  },
  correctAnswer: {
    color: '#22C55E',
  },
  footer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  attemptCount: {
    fontSize: 12,
    color: '#6B7280',
    fontWeight: '500',
  },
  viewDetailsContainer: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  viewDetailsText: {
    fontSize: 12,
    color: '#3B82F6',
    fontWeight: '500',
    marginRight: 4,
  },
  chevron: {
    fontSize: 16,
    color: '#3B82F6',
    fontWeight: '600',
  },
}); 