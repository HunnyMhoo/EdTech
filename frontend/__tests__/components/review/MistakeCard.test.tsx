import React from 'react';
import { create } from 'react-test-renderer';
import { TouchableOpacity } from 'react-native';
import { MistakeCard } from '../../../src/features/review/components/MistakeCard';
import { ReviewMistakeItem } from '../../../src/features/review/api/reviewMistakesApi';

describe('MistakeCard', () => {
  const mockMistake: ReviewMistakeItem = {
    question_id: 'Q1',
    question_text: 'What is 2+2?',
    skill_area: 'Mathematics',
    difficulty_level: 1,
    choices: [
      { id: 'A', text: '3' },
      { id: 'B', text: '4' },
      { id: 'C', text: '5' }
    ],
    user_answer_id: 'A',
    user_answer_text: '3',
    correct_answer_id: 'B',
    correct_answer_text: '4',
    explanation: 'The correct answer is 4.',
    mission_date: '2024-01-15',
    mission_completion_date: '2024-01-15T08:30:00Z',
    attempt_count: 2
  };

  const mockOnPress = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly without crashing', () => {
    const component = create(
      <MistakeCard mistake={mockMistake} onPress={mockOnPress} />
    );
    expect(component).toBeTruthy();
  });

  it('calls onPress when TouchableOpacity is pressed', () => {
    const component = create(
      <MistakeCard mistake={mockMistake} onPress={mockOnPress} />
    );
    
    const touchable = component.root.findByType(TouchableOpacity);
    touchable.props.onPress();
    
    expect(mockOnPress).toHaveBeenCalledTimes(1);
  });

  it('renders with showDate false', () => {
    const component = create(
      <MistakeCard 
        mistake={mockMistake} 
        onPress={mockOnPress} 
        showDate={false} 
      />
    );
    expect(component).toBeTruthy();
  });

  it('renders with compact mode', () => {
    const component = create(
      <MistakeCard 
        mistake={mockMistake} 
        onPress={mockOnPress} 
        compact={true} 
      />
    );
    expect(component).toBeTruthy();
  });

  it('handles different difficulty levels', () => {
    const testCases = [1, 2, 3, 4];
    
    testCases.forEach((level) => {
      const mistakeWithLevel = { ...mockMistake, difficulty_level: level };
      const component = create(
        <MistakeCard mistake={mistakeWithLevel} onPress={mockOnPress} />
      );
      expect(component).toBeTruthy();
    });
  });

  it('handles single attempt count', () => {
    const singleAttemptMistake = { ...mockMistake, attempt_count: 1 };
    const component = create(
      <MistakeCard mistake={singleAttemptMistake} onPress={mockOnPress} />
    );
    expect(component).toBeTruthy();
  });

  it('matches snapshot', () => {
    const component = create(
      <MistakeCard mistake={mockMistake} onPress={mockOnPress} />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
}); 