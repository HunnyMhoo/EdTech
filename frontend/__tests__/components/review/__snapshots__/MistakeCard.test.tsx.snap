// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MistakeCard matches snapshot 1`] = `null`;
